File: apps/ios/Janark/UI/Components/EngageBar.swift
import SwiftUI
import UIKit

struct EngageBar: View {
    @EnvironmentObject private var app: AppContainer
    @Environment(\.janarkColors) private var c

    let targetType: String
    let targetId: String
    let sharePath: String
    let shareTitle: String?
    var compact: Bool = true
    var onOpenComments: (() -> Void)? = nil
    var showComments: Bool = false

    @State private var counts: EngageCountsDto?
    @State private var error: String?
    @State private var flagOpen = false
    @State private var commentsOpen = false
    @State private var sharePayload: SharePayload?

    var body: some View {
        VStack(alignment: .leading, spacing: 8) {
            HStack {
                HStack(spacing: 4) {
                    Button { vote("upvote") } label: {
                        Image(systemName: counts?.myVote == 1 ? "hand.thumbsup.fill" : "hand.thumbsup")
                            .foregroundStyle(counts?.myVote == 1 ? c.amber : c.muted)
                    }
                    Text(indianCount(counts?.upvotes ?? 0))
                        .font(.subheadline)
                        .foregroundStyle(c.foreground)
                    Button { vote("downvote") } label: {
                        Image(systemName: counts?.myVote == -1 ? "hand.thumbsdown.fill" : "hand.thumbsdown")
                            .foregroundStyle(counts?.myVote == -1 ? c.danger : c.muted)
                    }
                    Button {
                        if showComments {
                            // Comments already inline — avoid a duplicate sheet
                            return
                        }
                        if let onOpenComments {
                            onOpenComments()
                        } else {
                            commentsOpen = true
                        }
                    } label: {
                        Image(systemName: "bubble.right")
                            .foregroundStyle(c.muted)
                    }
                    Text(indianCount(counts?.commentCount ?? 0))
                        .font(.subheadline)
                        .foregroundStyle(c.muted)
                }
                Spacer()
                HStack(spacing: 4) {
                    Button { share() } label: {
                        Image(systemName: "square.and.arrow.up")
                            .foregroundStyle(c.muted)
                    }
                    Button {
                        gated("flag this") { flagOpen = true }
                    } label: {
                        Image(systemName: "flag")
                            .foregroundStyle(c.muted)
                    }
                }
            }
            .buttonStyle(.plain)

            if let error {
                Text(error)
                    .font(.caption)
                    .foregroundStyle(c.danger)
            }

            if showComments {
                CommentThread(targetType: targetType, targetId: targetId)
            }
        }
        .task(id: "\(targetType)-\(targetId)") { await load() }
        .sheet(isPresented: $flagOpen) {
            FlagSheet(targetType: targetType, targetId: targetId) { flagOpen = false }
        }
        .sheet(isPresented: $commentsOpen) {
            NavigationStack {
                ScrollView {
                    CommentThread(targetType: targetType, targetId: targetId)
                        .padding()
                }
                .navigationTitle("Comments")
                .navigationBarTitleDisplayMode(.inline)
                .toolbar {
                    ToolbarItem(placement: .cancellationAction) {
                        Button("Close") { commentsOpen = false }
                    }
                }
            }
        }
        .sheet(item: $sharePayload) { payload in
            ActivityView(activityItems: [payload.text])
        }
    }

    private func load() async {
        switch await app.repository.engage(targetType: targetType, targetId: targetId) {
        case .ok(let value): counts = value
        case .err(let message, _, _): error = message
        }
    }

    private func gated(_ reason: String, _ block: @escaping () -> Void) {
        if app.session.uiSession.authenticated {
            block()
        } else {
            app.auth.require(reason, onSuccess: block)
        }
    }

    private func vote(_ choice: String) {
        gated("vote") {
            Task {
                switch await app.repository.voteEngage(targetType: targetType, targetId: targetId, choice: choice) {
                case .ok(let value):
                    counts = value
                case .err(let message, let unauthorized, _):
                    if unauthorized {
                        app.auth.require("vote") { vote(choice) }
                    } else {
                        error = message
                    }
                }
            }
        }
    }

    private func share() {
        gated("share") {
            Task {
                _ = await app.repository.trackShare(platform: "native", path: sharePath, title: shareTitle)
                let base = app.session.currentApiBase()
                let absolute: String = {
                    if sharePath.hasPrefix("http://") || sharePath.hasPrefix("https://") {
                        return sharePath
                    }
                    let path = sharePath.hasPrefix("/") ? sharePath : "/\(sharePath)"
                    return base + path
                }()
                let text = [shareTitle, absolute].compactMap { $0 }.filter { !$0.isEmpty }.joined(separator: "\n")
                sharePayload = SharePayload(text: text.isEmpty ? "Janark" : text)
            }
        }
    }
}

private struct SharePayload: Identifiable {
    let id = UUID()
    let text: String
}

struct ActivityView: UIViewControllerRepresentable {
    let activityItems: [Any]

    func makeUIViewController(context: Context) -> UIActivityViewController {
        UIActivityViewController(activityItems: activityItems, applicationActivities: nil)
    }

    func updateUIViewController(_ uiViewController: UIActivityViewController, context: Context) {}
}

File: apps/android/app/src/main/java/org/janark/app/ui/components/EngageBar.kt
package org.janark.app.ui.components

import android.content.Intent
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.outlined.ChatBubbleOutline
import androidx.compose.material.icons.outlined.Flag
import androidx.compose.material.icons.outlined.IosShare
import androidx.compose.material.icons.outlined.ThumbDown
import androidx.compose.material.icons.outlined.ThumbUp
import androidx.compose.material.icons.filled.ThumbDown
import androidx.compose.material.icons.filled.ThumbUp
import androidx.compose.material3.Icon
import androidx.compose.material3.IconButton
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.runtime.collectAsState
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.rememberCoroutineScope
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.platform.LocalContext
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import kotlinx.coroutines.launch
import org.janark.app.core.Outcome
import org.janark.app.core.indianCount
import org.janark.app.data.net.EngageCountsDto
import org.janark.app.ui.LocalApp
import org.janark.app.ui.theme.JanarkTheme
import org.janark.app.ui.theme.toC

@Composable
fun EngageBar(
    targetType: String,
    targetId: String,
    sharePath: String,
    shareTitle: String?,
    compact: Boolean = true,
    onOpenComments: (() -> Unit)? = null,
    showComments: Boolean = !compact,
) {
    val app = LocalApp.current
    val session by app.session.uiSession.collectAsState(
        initial = org.janark.app.data.session.UiSession(),
    )
    var counts by remember(targetType, targetId) { mutableStateOf<EngageCountsDto?>(null) }
    var error by remember { mutableStateOf<String?>(null) }
    var flagOpen by remember { mutableStateOf(false) }
    var commentsOpen by remember { mutableStateOf(false) }
    val scope = rememberCoroutineScope()
    val ctx = LocalContext.current
    val c = JanarkTheme.colors

    fun load() {
        scope.launch {
            when (val r = app.repository.engage(targetType, targetId)) {
                is Outcome.Ok -> counts = r.value
                is Outcome.Err -> error = r.message
            }
        }
    }

    LaunchedEffect(targetType, targetId) { load() }

    fun gated(reason: String, block: suspend () -> Unit) {
        if (!session.authenticated) {
            app.auth.require(reason) { scope.launch { block() } }
        } else {
            scope.launch { block() }
        }
    }

    fun vote(choice: String) {
        gated("vote") {
            when (val r = app.repository.voteEngage(targetType, targetId, choice)) {
                is Outcome.Ok -> counts = r.value
                is Outcome.Err -> {
                    if (r.unauthorized) app.auth.require("vote") { vote(choice) }
                    else error = r.message
                }
            }
        }
    }

    Column(Modifier.fillMaxWidth()) {
        Row(
            Modifier.fillMaxWidth(),
            verticalAlignment = Alignment.CenterVertically,
            horizontalArrangement = Arrangement.SpaceBetween,
        ) {
            Row(verticalAlignment = Alignment.CenterVertically) {
                IconButton(onClick = { vote("upvote") }) {
                    Icon(
                        if (counts?.myVote == 1) Icons.Filled.ThumbUp else Icons.Outlined.ThumbUp,
                        contentDescription = "Upvote",
                        tint = if (counts?.myVote == 1) c.amber.toC() else c.muted.toC(),
                    )
                }
                Text(indianCount(counts?.upvotes ?: 0), color = c.foreground.toC(), fontSize = 13.sp)
                IconButton(onClick = { vote("downvote") }) {
                    Icon(
                        if (counts?.myVote == -1) Icons.Filled.ThumbDown else Icons.Outlined.ThumbDown,
                        contentDescription = "Downvote",
                        tint = if (counts?.myVote == -1) c.danger.toC() else c.muted.toC(),
                    )
                }
                IconButton(onClick = {
                    if (onOpenComments != null) onOpenComments.invoke()
                    else commentsOpen = !commentsOpen
                }) {
                    Icon(Icons.Outlined.ChatBubbleOutline, contentDescription = "Comments", tint = c.muted.toC())
                }
                Text(indianCount(counts?.commentCount ?: 0), color = c.muted.toC(), fontSize = 13.sp)
            }
            Row {
                IconButton(onClick = {
                    gated("share") {
                        app.repository.trackShare("native", sharePath, shareTitle)
                        val base = app.session.currentApiBase()
                        val absolute = when {
                            sharePath.startsWith("http://") || sharePath.startsWith("https://") -> sharePath
                            sharePath.startsWith("/") -> "$base$sharePath"
                            else -> "$base/$sharePath"
                        }
                        val body = listOfNotNull(shareTitle, absolute)
                            .filter { it.isNotBlank() }
                            .joinToString("\n")
                        val send = Intent(Intent.ACTION_SEND).apply {
                            type = "text/plain"
                            putExtra(Intent.EXTRA_SUBJECT, shareTitle ?: "Janark")
                            putExtra(Intent.EXTRA_TEXT, body.ifBlank { "Janark" })
                        }
                        ctx.startActivity(Intent.createChooser(send, "Share"))
                    }
                }) {
                    Icon(Icons.Outlined.IosShare, contentDescription = "Share", tint = c.muted.toC())
                }
                IconButton(onClick = {
                    gated("flag this") { flagOpen = true }
                }) {
                    Icon(Icons.Outlined.Flag, contentDescription = "Report", tint = c.muted.toC())
                }
            }
        }
        error?.let { Text(it, color = c.danger.toC(), fontSize = 12.sp) }
        if (showComments || commentsOpen) {
            CommentThread(targetType, targetId)
        }
        if (flagOpen) {
            FlagSheet(targetType, targetId, onDismiss = { flagOpen = false })
        }
    }
}
